src/integer/mat_z/tensor.rs

```diff
@@ -0,0 +1,214 @@
+// Copyright © 2023 Marvin Beckmann
+//
+// This file is part of qFALL-math.
+//
+// qFALL-math is free software: you can redistribute it and/or modify it under
+// the terms of the Mozilla Public License Version 2.0 as published by the
+// Mozilla Foundation. See <https://mozilla.org/en-US/MPL/2.0/>.
+
+//! This module contains the implementation of the `tensor` product.
+
+use super::MatZ;
+use crate::traits::{GetNumColumns, GetNumRows, Tensor};
+use flint_sys::{
+    fmpz::{fmpz, fmpz_is_zero, fmpz_mul},
+    fmpz_mat::fmpz_mat_entry,
+};
+
+impl Tensor for MatZ {
+    /// Computes the tensor product of `self` with `other`
+    ///
+    /// Parameters:
+    /// - `other`: the value with which the tensor product is computed.
+    ///
+    /// Returns the tensor product of `self` with `other`.
+    ///
+    /// # Example
+    /// ```
+    /// use qfall_math::integer::MatZ;
+    /// use qfall_math::traits::Tensor;
+    /// use std::str::FromStr;
+    ///
+    /// let mat_1 = MatZ::from_str("[[1, 0, 0],[0, 1, 0],[0, 0, 1]]").unwrap();
+    /// let mat_2 = MatZ::from_str("[[1, 2, 1],[3, 4, 1]]").unwrap();
+    ///
+    /// let mat_3 = mat_1.tensor(&mat_2);
+    /// ```
+    fn tensor(&self, other: &Self) -> Self {
+        // we can unwrap since we know that the dimensions are positive
+        // does not have to be mutable, since the pointers themselves
+        // do not change, only the content behind the pointers.
+        let out = MatZ::new(
+            self.get_num_rows() * other.get_num_rows(),
+            self.get_num_columns() * other.get_num_columns(),
+        )
+        .unwrap();
+
+        for i in 0..self.get_num_rows() {
+            for j in 0..self.get_num_columns() {
+                let entry = unsafe { fmpz_mat_entry(&self.matrix, i, j) };
+
+                if unsafe { 1 != fmpz_is_zero(entry) } {
+                    set_partial_tensor_fmpz(&out, i, j, entry, other)
+                }
+            }
+        }
+
+        out
+    }
+}
+
+/// This function sets the entries of the provided matrix according to the tensor product.
+/// It assumes that the matrix is correctly instantiated with zeros.
+///
+/// Sets the entries
+/// `[i*rows_other, j*columns_other]` up till `[i*(row_other +1), j*(columns_other + 1)]`
+/// according to the definition of the tensor product
+///
+/// Parameters:
+/// - `out`: the matrix in which the result is entered
+/// (**Warning**: even though `out` is not mutable, the subpart of the matrix is still
+/// set since the content behind the pointers changed and not the pointers themselves)
+/// - `i`: defines the row from which `entry` is taken
+/// - `j`: defines the column from which `entry` is taken
```

Comment:
Please describe these a little more precise, e.g. `defines the leftmost row of the set window`